test/spec/runners/ssz_static.ex

```diff
@@ -73,28 +75,102 @@ defmodule SszStaticTestRunner do
     assert {:ok, decompressed} = :snappyer.decompress(compressed)
 
-    expected =
+    sanitize_yaml =
       YamlElixir.read_from_file!(case_dir <> "/value.yaml")
       |> SpecTestUtils.sanitize_yaml()
-      |> SpecTestUtils.sanitize_ssz(schema)
+
+    expected_sanitized =
+      SpecTestUtils.sanitize_ssz(sanitize_yaml, schema)
 
     %{"root" => expected_root} = YamlElixir.read_from_file!(case_dir <> "/roots.yaml")
     expected_root = expected_root |> SpecTestUtils.sanitize_yaml()
 
-    assert_ssz(schema, decompressed, expected, expected_root)
+    assert_ssz(schema, decompressed, expected_sanitized, sanitize_yaml, expected_root)
   end
 
-  defp assert_ssz(schema, real_serialized, real_deserialized, expected_root) do
+  defp assert_ssz(
+         schema,
+         real_serialized,
+         real_deserialized_sanitized,
+         real_deserialized,
+         _expected_root
+       ) do
     {:ok, deserialized} = SszEx.decode(real_serialized, schema)
-    assert Diff.diff(deserialized, real_deserialized) == :unchanged
-    {:ok, serialized} = SszEx.encode(real_deserialized, schema)
+    assert Diff.diff(deserialized, real_deserialized_sanitized) == :unchanged
+
+    {:ok, deserialized_old} = Ssz.from_ssz(real_serialized, schema)
+    real_deserialized = to_struct_checked(deserialized_old, real_deserialized)
+
+    assert Diff.diff(
+             deserialized,
+             SpecTestUtils.sanitize_ssz(from_deep_struct_to_map(real_deserialized), schema)
+           ) == :unchanged
+
+    {:ok, serialized} = SszEx.encode(real_deserialized_sanitized, schema)
     assert serialized == real_serialized
 
-    root = SszEx.hash_tree_root!(real_deserialized, schema)
-    assert root == expected_root
+    {:ok, serialized_old} = Ssz.to_ssz(real_deserialized)
+    assert Diff.diff(serialized, serialized_old) == :unchanged
+
+    # TODO enable when hash_tree_root supports supports new schema types :byte_list
+    # root = SszEx.hash_tree_root!(real_deserialized, schema)
+    # assert root == expected_root
+  end
+
+  def from_deep_struct_to_map(%{} = map), do: convert(map)
+
+  defp convert(vec(_) = actual) do
+    Aja.Enum.to_list(actual) |> convert()
+  end
+
+  defp convert(data) when is_struct(data) do
+    data |> Map.from_struct() |> convert()
   end
 
+  defp convert(data) when is_map(data) do
+    for {key, value} <- data, reduce: %{} do
+      acc ->
+        case key do
+          :__meta__ ->
+            acc
+
+          other ->
+            Map.put(acc, other, convert(value))
+        end
+    end
+  end
+
+  defp convert(data) when is_list(data) do
+    for element <- data, reduce: [] do
+      acc ->
+        [convert(element) | acc]
+    end
+    |> Enum.reverse()
+  end
+
+  defp convert(other), do: other
+
   defp parse_type(%SpecTestCase{handler: handler}) do
     Module.concat(Types, handler)
   end
+
+  defp to_struct_checked(actual, expected) when is_list(actual) and is_list(expected) do
```

Comment:
What does this function do exactly? The name is not very clear. We might want to add it in a comment. From what I'm guessing:

- If it's a vec or a list it just follows recursively.
- If it's a base element, it just discards `actual` and takes from `expected`.
- When receiving a struct and a map it uses the struct name of `actual` but the data from `expected`, effectively discarding `actual`.

If this is the case, it means that `real_deserialized` in the calling function is actually the parsed value from the yaml with the struct names of the ssz rust library? If that's the case, why do we do this? We are not comparing the deserialized value from both libraries, we are doing extremely complex processing to compare the new deserialized value with the parsed value again. Shouldn't we take the actual values from `old_deserialized` and try to make the format compatible with the new one to compare instead?

Reply:
this method `to_struct_checked`  was already present with the NIF Ssz Rust :sweat_smile:  I removed it when I started to migrate the spec test to the Elixir Lib. Then, I added it again to try to compare the result of the two libraries

Reply:
From what I understand this method is similar to `sanitize_ssz` because it converts maps to the corresponding structs and also tries to fix some compatibilities between list and the Aja library.